syncmaster/scheduler/transfer_job_manager.py

```diff
@@ -0,0 +1,83 @@
+# SPDX-FileCopyrightText: 2023-2024 MTS PJSC
+# SPDX-License-Identifier: Apache-2.0
+import asyncio
+
+from apscheduler.schedulers.asyncio import AsyncIOScheduler
+from apscheduler.triggers.cron import CronTrigger
+from kombu.exceptions import KombuError
+
+from syncmaster.backend.services.unit_of_work import UnitOfWork
+from syncmaster.db.models import RunType, Status, Transfer
+from syncmaster.exceptions.run import CannotConnectToTaskQueueError
+from syncmaster.scheduler.transfer_fetcher import get_async_session
+from syncmaster.schemas.v1.connections.connection import ReadAuthDataSchema
+from syncmaster.settings import Settings
+from syncmaster.worker.config import celery
+
+
+class TransferJobManager:
+    def __init__(self, settings: Settings):
+        self.scheduler = AsyncIOScheduler(timezone=settings.TZ)
+        self.scheduler.add_jobstore("sqlalchemy", url=settings.database.sync_url)
+        self.settings = settings
+
+    def update_jobs(self, transfers: list[Transfer]) -> None:
+        for transfer in transfers:
+            job_id = str(transfer.id)
+            existing_job = self.scheduler.get_job(job_id)
+
+            if not transfer.is_scheduled or transfer.is_deleted:
+                if existing_job:
+                    self.scheduler.remove_job(job_id)
+                continue
+
+            if existing_job:
+                self.scheduler.modify_job(
+                    job_id=job_id,
+                    trigger=CronTrigger.from_crontab(transfer.schedule),
+                    misfire_grace_time=self.settings.SCHEDULER_MISFIRE_GRACE_TIME,
+                    args=(transfer.id,),
+                )
+            else:
+                self.scheduler.add_job(
+                    func=TransferJobManager.send_job_to_celery,
+                    id=job_id,
+                    trigger=CronTrigger.from_crontab(transfer.schedule),
+                    misfire_grace_time=self.settings.SCHEDULER_MISFIRE_GRACE_TIME,
+                    args=(transfer.id,),
+                )
+
+    @staticmethod
+    async def send_job_to_celery(transfer_id: int) -> None:
+        """
+        Do not pass additional arguments like settings,
+        otherwise they will be serialized in jobs table.
+        """
+        settings = Settings()
```

Comment:
Why settings object is created again, instead of reusing self attribute?

Reply:
Apscheduler requires that the function passed to add_job() be serializable. If I were to pass an instance method of a class, that method would be bound to a specific object. As a result, фpscheduler cannot serialize or reliably store a reference to such a method, the weak reference problem appears. Therefore, I pass a static method to the job, since such functions are not object-specific and can be serialized.